File: sources/wasm-gc-teavm/js/platformRuntime.js
/*
 * Copyright (c) 2024 lax1dude. All Rights Reserved.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 * 
 */

const platfRuntimeName = "platformRuntime";

var allowImmediateContinue = false;
const immediateContinueChannel = new MessageChannel();
var immediateContinueHandler = null;

immediateContinueChannel.port2.addEventListener("message", function(evt) {
	immediateContinueHandler();
});

async function initializePlatfRuntime() {
	immediateContinueChannel.port1.start();
	immediateContinueChannel.port2.start();

	immediateContinueHandler = function() {
		immediateContinueHandler = null;
	};

	immediateContinueChannel.port1.postMessage(0);

	if(immediateContinueHandler) {
		await new Promise(function(resolve) {
			setTimeout(function() {
				if(!immediateContinueHandler) {
					allowImmediateContinue = true;
				}else {
					eagError("Immediate continue hack is not supported");
				}
				resolve();
			}, 25);
		});
	}else {
		eagError("Immediate continue hack is not supported");
	}
}

/**
 * @return {HTMLElement}
 */
eagruntimeImpl.platformRuntime["getRootElement"] = function() {
	return rootElement;
};

/**
 * @return {HTMLElement}
 */
eagruntimeImpl.platformRuntime["getParentElement"] = function() {
	return parentElement;
};

/**
 * @return {HTMLCanvasElement}
 */
eagruntimeImpl.platformRuntime["getCanvasElement"] = function() {
	return canvasElement;
};

/**
 * @return {Object}
 */
eagruntimeImpl.platformRuntime["getEaglercraftXOpts"] = function() {
	return eaglercraftXOpts;
};

eagruntimeImpl.platformRuntime["getEventCount"] = mainEventQueue.getLength.bind(mainEventQueue);

eagruntimeImpl.platformRuntime["getNextEvent"] = mainEventQueue.shift.bind(mainEventQueue);

const EVENT_RUNTIME_ASYNC_DOWNLOAD = 0;

/**
 * @param {string} uri
 * @param {number} forceCache
 * @param {number} id
 */
eagruntimeImpl.platformRuntime["queueAsyncDownload"] = function(uri, forceCache, id) {
	try {
		fetch(uri, /** @type {!RequestInit} */ ({
			"cache": forceCache ? "force-cache" : "no-store",
			"mode": "cors"
		})).then(function(res) {
			return res.arrayBuffer();
		}).then(function(arr) {
			pushEvent(EVENT_TYPE_RUNTIME, EVENT_RUNTIME_ASYNC_DOWNLOAD, {
				"requestId": id,
				"arrayBuffer": arr
			});
		}).catch(function(err) {
			eagError("Failed to complete async download: {}", uri);
			eagStackTrace(ERROR, "Exception Caught", /** @type {Error} */ (err));
			pushEvent(EVENT_TYPE_RUNTIME, EVENT_RUNTIME_ASYNC_DOWNLOAD, {
				"requestId": id,
				"arrayBuffer": null
			});
		});
	}catch(/** Error */ ex) {
		eagError("Failed to fetch: {}", uri);
		eagStackTrace(ERROR, "Exception Caught", ex);
		pushEvent(EVENT_TYPE_RUNTIME, EVENT_RUNTIME_ASYNC_DOWNLOAD, {
			"requestId": id,
			"arrayBuffer": null
		});
	}
};

/**
 * @param {string} uri
 * @param {number} forceCache
 * @return {Promise}
 */
function downloadImpl(uri, forceCache) {
	return new Promise(function(resolve) {
		try {
			fetch(uri, /** @type {!RequestInit} */ ({
				"cache": forceCache ? "force-cache" : "no-store",
				"mode": "cors"
			})).then(function(res) {
				return res.arrayBuffer();
			}).then(function(arr) {
				resolve(arr);
			}).catch(function(err) {
				eagError("Failed to complete download: {}", uri);
				eagStackTrace(ERROR, "Exception Caught", /** @type {Error} */ (err));
				resolve(null);
			});
		}catch(/** Error */ ex) {
			eagError("Failed to fetch: {}", uri);
			eagStackTrace(ERROR, "Exception Caught", ex);
			resolve(null);
		}
	});
}

eagruntimeImpl.platformRuntime["download"] = new WebAssembly.Suspending(downloadImpl);

/**
 * @param {string} crashDump
 */
eagruntimeImpl.platformRuntime["writeCrashReport"] = function(crashDump) {
	displayCrashReport(crashDump, false);
};

eagruntimeImpl.platformRuntime["steadyTimeMillis"] = performance.now.bind(performance);

/**
 * @param {number} millis
 * @return {Promise}
 */
function sleepImpl(millis) {
	return new Promise(function(resolve) {
		setTimeout(resolve, millis);
	});
}

eagruntimeImpl.platformRuntime["sleep"] = new WebAssembly.Suspending(sleepImpl);

function immediateContinueResolver(resolve) {
	if(allowImmediateContinue) {
		immediateContinueHandler = resolve;
		immediateContinueChannel.port1.postMessage(0);
	}else {
		setTimeout(resolve, 0);
	}
}

/**
 * @return {Promise}
 */
function immediateContinueImpl() {
	return new Promise(immediateContinueResolver);
}

/**
 * @return {Promise}
 */
function swapDelayImpl() {
	if(!runtimeOpts.useDelayOnSwap) {
		return immediateContinueImpl();
	}else {
		return sleepImpl(0);
	}
}

eagruntimeImpl.platformRuntime["immediateContinue"] = new WebAssembly.Suspending(immediateContinueImpl);

/**
 * @param {number} id
 * @param {string} str
 */
eagruntimeImpl.platformRuntime["setCrashReportString"] = function(id, str) {
	crashReportStrings[id] = str;
};

File: sources/platform-api/java/net/lax1dude/eaglercraft/v1_8/internal/PlatformRuntime.java
/*
 * Copyright (c) 2025 lax1dude. All Rights Reserved.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 * 
 */

package net.lax1dude.eaglercraft.v1_8.internal;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.function.Consumer;

import net.lax1dude.eaglercraft.v1_8.internal.buffer.ByteBuffer;
import net.lax1dude.eaglercraft.v1_8.internal.buffer.FloatBuffer;
import net.lax1dude.eaglercraft.v1_8.internal.buffer.IntBuffer;

public class PlatformRuntime {

	public static native void create();

	public static native void destroy();

	public static native EnumPlatformType getPlatformType();

	public static native EnumPlatformAgent getPlatformAgent();

	public static native String getUserAgentString();

	public static native EnumPlatformOS getPlatformOS();

	public static native boolean requireSSL();

	public static native boolean isOfflineDownloadURL();

	public static native EnumPlatformANGLE getPlatformANGLE();

	public static native String getGLVersion();

	public static native String getGLRenderer();

	public static native ByteBuffer allocateByteBuffer(int length);

	public static native IntBuffer allocateIntBuffer(int length);

	public static native FloatBuffer allocateFloatBuffer(int length);

	public static native ByteBuffer castPrimitiveByteArray(byte[] array);

	public static native IntBuffer castPrimitiveIntArray(int[] array);

	public static native FloatBuffer castPrimitiveFloatArray(float[] array);

	public static native byte[] castNativeByteBuffer(ByteBuffer buffer);

	public static native int[] castNativeIntBuffer(IntBuffer buffer);

	public static native float[] castNativeFloatBuffer(FloatBuffer buffer);

	public static native void freeByteBuffer(ByteBuffer byteBuffer);

	public static native void freeIntBuffer(IntBuffer intBuffer);

	public static native void freeFloatBuffer(FloatBuffer floatBuffer);

	public static native void downloadRemoteURIByteArray(String assetPackageURI, final Consumer<byte[]> cb);

	public static native void downloadRemoteURIByteArray(String assetPackageURI, boolean useCache,
			final Consumer<byte[]> cb);

	public static native byte[] downloadRemoteURIByteArray(String assetPackageURI);

	public static native byte[] downloadRemoteURIByteArray(String assetPackageURI, boolean forceCache);

	public static native boolean isDebugRuntime();

	public static native void writeCrashReport(String crashDump);

	public static native void showContextLostScreen(String crashDump);

	public static native void getStackTrace(Throwable t, Consumer<String> ret);

	public static native boolean printJSExceptionIfBrowser(Throwable t);

	public static native String currentThreadName();

	public static native void setThreadName(String string);

	public static native long maxMemory();

	public static native long totalMemory();

	public static native long freeMemory();

	public static native String getCallingClass(int i);

	public static native long randomSeed();

	public static native void exit();

	public static native IClientConfigAdapter getClientConfigAdapter();

	public static native long steadyTimeMillis();

	public static native long nanoTime();

	public static native void sleep(int millis);

	public static native void immediateContinue();

	public static native boolean immediateContinueSupported();

	public static native void postCreate();

	public static native void setDisplayBootMenuNextRefresh(boolean en);

	public static native OutputStream newDeflaterOutputStream(OutputStream os) throws IOException;

	public static native int deflateFull(byte[] input, int inputOff, int inputLen, byte[] output, int outputOff,
			int outputLen) throws IOException;

	public static native OutputStream newGZIPOutputStream(OutputStream os) throws IOException;

	public static native InputStream newInflaterInputStream(InputStream is) throws IOException;

	public static native int inflateFull(byte[] input, int inputOff, int inputLen, byte[] output, int outputOff,
			int outputLen) throws IOException;

	public static native InputStream newGZIPInputStream(InputStream is) throws IOException;

}